### ISSUE TITLE
Thumbnails do not show up on Gnome anymore

### ISSUE THREAD

**Issue by ysooqe** [xournalpp/xournalpp#1612]
**Affects versions :**
 - OS: Linux
 - Gnome (X11)
 - Version of Xournal++ : 1.0.14 and 1.0.16

**Description**
Since 1.0.14, Thumbnails for .xopp files do not show up in the file manager (Nautilus) anymore

**To Reproduce**
1. Open the file browser and look for a .xopp file.

**Expected behavior**
The .xopp file has a thumbnail that shows the contents of the first page of the file.

**Actual behavior**
It uses the generic icon of Xournalpp instead of thumbnails.

**Comment by LittleHuba** [comment]
Did you install one of the prior versions manually using `make install`?

**Comment by ysooqe** [comment]
> Did you install one of the prior versions manually using `make install`?

No, I have not done that.

**Comment by LittleHuba** [comment]
Please check out this comment anyway and check if those files exist:
https://github.com/xournalpp/xournalpp/issues/1442#issuecomment-524566511

**Comment by Rmano** [comment]
I have installed 1.0.16 (from git branch with `sudo make install`, and to be on the safe side I did: 
```bash
CMAKE_INSTALL_PREFIX=/usr/local
rm -f ${CMAKE_INSTALL_PREFIX}/share/icons/hicolor/scalable/apps/xournalpp.svg
rm -f ${CMAKE_INSTALL_PREFIX}/share/mime/packages/xournalpp.xml
rm -f ${CMAKE_INSTALL_PREFIX}/share/applications/xournalpp.desktop
rm -f ${CMAKE_INSTALL_PREFIX}/share/thumbnailers/xournalpp.thumbnailer
rm -f ${CMAKE_INSTALL_PREFIX}/share/metainfo/xournalpp.appdata.xml
``` 
(as root) before installing, and still I have this: 
![image](https://user-images.githubusercontent.com/6414907/69073487-d87b8080-0a2d-11ea-8c29-be82a1f3ca35.png)

**Comment by LittleHuba** [comment]
@Rmano you have to run:
```bash
./cmake/postinst configure
```

**Comment by Rmano** [comment]
Done; no error. I removed the thumbnail cache (`rm -rf ~/.cache/thumbnails`) and still I have the same behavior. 
I will investigate a bit more...

**Comment by ysooqe** [comment]
> Please check out this comment anyway and check if those files exist:
> [#1442 (comment)](https://github.com/xournalpp/xournalpp/issues/1442#issuecomment-524566511)

I checked and all those files exist.

**Comment by LittleHuba** [comment]
As mentioned in the comment. Please delete the files.

**Comment by ysooqe** [comment]
> As mentioned in the comment. Please delete the files.

Sorry, that was not very clear from my side:
I do have the files that SHOULD be there. None of the files are named `xournalpp` but `com.github.xournalpp.xournalpp` as they should be according to the linked comment.

**Comment by Technius** [comment]
Looks like thumbnail generation is failing. Here's a file I found in `~/.cache/thumbnails/fail/gnome-thumbnail-factory/` after saving a dummy file and opening the folder in a file manager (I deleted `~/.cache/thumbnails` before):
```plain
<89>PNG
^Z
^@^@^@^MIHDR^@^@^@^A^@^@^@^A^H^F^@^@^@^_^Uĉ^@^@^@^Dd<88>^@^@^@^_tEXtThumb::URI^@file:///tmp/foo.xopp<99><C2>L<EA>^@^@^@^WtEXtThumb::MTime^@1574923269<A6>
<A6>^@^@^@ tEXtSoftware^@GNOME::ThumbnailFactoryoDU<E0>^@^@^@^KIDA<99>c`^@^B^@^@^E^@^AbU2<88>^@^@^@^@IEND<AE>B`<82>

```

**Comment by Technius** [comment]
This appears to be [a bug in Nautilus](http://www.bernaerts-nicolas.fr/linux/74-ubuntu/360-ubuntu-nautilus-external-thumbnailer-failure).

I cleared the thumbnail cache and opened the containing folder with Nemo; the thumbnail displayed correctly.

**Comment by Rmano** [comment]
> This appears to be [a bug in Nautilus](http://www.bernaerts-nicolas.fr/linux/74-ubuntu/360-ubuntu-nautilus-external-thumbnailer-failure).
> 
> I cleared the thumbnail cache and opened the containing folder with Nemo; the thumbnail displayed correctly.

Hmm, I will check, but I do not think that this is the problem. I have the workaround described in the linked post installed since more than one year, and nevertheless the problem should have been fixed upstream (I'll try to dig out the links later).

Mind you, it could still be a Nautilus bug, but I don't think is this one.

**Comment by jvwowd** [comment]
This does not directly address your issue, but I thought I would mention it in case it is related. I found that when I export my Xournal++ file to .pdf, Nautilus does not show the thumbnail (just shows the general .pdf icon), but when I use the print menu and choose to print to a .pdf file, then Nautilus shows the thumbnail of that .pdf properly.

**Comment by robinp7720** [comment]
I believe I've discovered the source of the issue. Nautilus uses bwrap to isolate thumbnailers, which also means that only the thumbnailer application and the file to thumbnail exist in the view of the thumbnailer. The thumbnail cannot be generated because the thumbnailer cannot find the xournalpp icon to overlay onto the thumbnail. I've got it working here: https://github.com/robinp7720/xournalpp/commit/b1e0a64232071a2269294d2c8329ec6f26583c4c by removing the code to overlay the xournalpp icon

**Comment by Technius** [comment]
Awesome, thanks for troubleshooting the issue. I dug some more into this issue and found the root cause: the image loading code that I wrote is a hack that uses `gtk_init` and `gdk_pixbuf`. Pixbuf dynamically loads its loader libraries, causing the `bwrap: execvp /tmp/xournalpp-thumbnailer: No such file or directory` errors.

The thumbnailer uses GTK because I wanted to avoid a `librsvg` dependency...but GTK already depends on `librsvg` so we might as well change it to use `librsvg` only! I'll see if I can fix this tomorrow.

**Comment by Febbe** [comment]
In PR #2252 I removed gtk_init will this either fix this issue or make it worse?

**Comment by Technius** [comment]
#2252 is unrelated to this issue, which only affects the thumbnailer.

**Comment by Technius** [comment]
Can someone try #2738? The thumbnails should at least show up--not sure if the icon will show up anymore though.

**Comment by robinp7720** [comment]
> Can someone try #2738? The thumbnails should at least show up--not sure if the icon will show up anymore though.

I can confirm that #2738 fixes the issue on nautilus, it even shows the xournallpp icon!

![xournalp_thumbnailer_working](https://user-images.githubusercontent.com/10325589/106721550-1352df80-6605-11eb-8fb1-6c100986b59d.png)

However, in some cases it seams that the thumbnail isn't being rendered. This was however an issue before #2738 with file managers that don't use bwrap. I'll send an offending file if desired.

![xournalpp_thumbnailer_notworking](https://user-images.githubusercontent.com/10325589/106721555-14840c80-6605-11eb-893d-14ebad7675df.png)

**Comment by Technius** [comment]
That's great. I think I'll leave the other issue to be fixed some other time.

**Comment by rolandlo** [comment]
Under Ubuntu 20.10 I only got the thumbnailer working after wrapping `bwrap` as described in [this solution](https://askubuntu.com/questions/1279091/nautilus-thumbnailer-for-ms-office-documents-in-ubuntu-20)

**Comment by moorepants** [comment]
I upgraded from Ubuntu 20.04 to 22.04 and am using the xournal++ snap package, now thumbnails display as gzip:
![image](https://user-images.githubusercontent.com/276007/172054873-b39960eb-dd9d-446a-b4a9-fbc5f30962b2.png)

**Comment by Technius** [comment]
@moorepants Does this problem also occur if you use the `xournalpp` deb package available in the official Ubuntu 22.04 repository instead of the snap?

**Comment by moorepants** [comment]
> @moorepants Does this problem also occur if you use the xournalpp deb package available in the official Ubuntu 22.04 repository instead of the snap?

I don't know. I'll have to switch and see. I was using the snap in 20.04 to have the latest version and the thumbnails worked then.

**Comment by moorepants** [comment]
The thumbnails work as expected with the deb package included with Ubuntu 22.04.
